packages/airnode-protocol/contracts/rrp/authorizers/admin/AirnodeTokenLock.sol

```diff
@@ -0,0 +1,495 @@
+    function setAirnodeFeeRegistry(address _airnodeFeeRegistry)
+        external
+        override
+    {
+        require(
+            hasCoefficientAndRegistrySetterRoleOrIsManager(msg.sender),
+            ERROR_NOT_COEFFICIENT_AND_REGISTRY_SETTER
+        );
+        require(_airnodeFeeRegistry != address(0), ERROR_ZERO_ADDRESS);
+        airnodeFeeRegistry = _airnodeFeeRegistry;
+        emit SetAirnodeFeeRegistry(_airnodeFeeRegistry, msg.sender);
+    }
+
+    /// @notice Called by a oracle address setter to set the status of an oracle address
+    /// @param _oracle The address of the oracle that can update the price
+    /// @param _status The status to be set
+    function setOracle(address _oracle, bool _status) external override {
+        require(
+            hasOracleAddressSetterRoleOrIsManager(msg.sender),
+            ERROR_NOT_ORACLE_ADDRESS_SETTER
+        );
+        require(_oracle != address(0), ERROR_ZERO_ADDRESS);
+        isOracle[_oracle] = _status;
+        emit SetOracle(_oracle, _status, msg.sender);
+    }
+
+    /// @notice Called by an oracle to set the price of API3
+    /// @param _price The price of API3 in USD
+    function setAPI3Price(uint256 _price) external override {
+        require(isOracle[msg.sender], ERROR_NOT_ORACLE);
+        require(_price != 0, ERROR_ZERO_AMOUNT);
+        api3PriceInUsd = _price;
+        emit SetAPI3Price(_price, msg.sender);
+    }
+
+    /// @notice Called by a coefficient and registry setter to set the multiplier coefficient
+    /// @param _multiplierCoefficient The multiplier coefficeint
+    function setMultiplierCoefficient(uint256 _multiplierCoefficient)
+        external
+        override
+    {
+        require(
+            hasCoefficientAndRegistrySetterRoleOrIsManager(msg.sender),
+            ERROR_NOT_COEFFICIENT_AND_REGISTRY_SETTER
+        );
+        require(_multiplierCoefficient != 0, ERROR_ZERO_AMOUNT);
+        multiplierCoefficient = _multiplierCoefficient;
+        emit SetMultiplierCoefficient(_multiplierCoefficient, msg.sender);
+    }
+
+    /// @notice Called by a opt status setter to set the opt status of an airnode
+    /// @param _airnode The airnode address
+    /// @param _status The Opted status for the airnode
+    function setOptInStatus(address _airnode, bool _status) external override {
+        require(
+            hasOptStatusSetterRoleOrIsManager(msg.sender),
+            ERROR_NOT_OPT_STATUS_SETTER
+        );
+        require(_airnode != address(0), ERROR_ZERO_ADDRESS);
+        airnodeOptInStatus[_airnode] = _status;
+        emit SetOptInStatus(_airnode, _status, msg.sender);
+    }
+
+    /// @notice Called by the airnode to set the opt status for itself
+    /// @param _airnode The airnode address
+    /// @param _status The Opted status for the airnode
+    function setSelfOptOutStatus(address _airnode, bool _status)
+        external
+        override
+    {
+        require(msg.sender == _airnode, ERROR_NOT_AIRNODE);
```

Comment:
This check is a little bit weird isn't it? 

Maybe what I would do instead is fetch the `airnodeOptInStatus` by `msg.sender` and check that is `true` since that is a requirement.

Reply:
both of these statuses are mutually exclusive, if an airnode is not opted in by an admin , we still want to give the airnode the ability to opt out incase they are opted in without consent.  Only the airnode needs to be able to set this status so the require makes sense or is there anything else I am missing?

Reply:
> if an airnode is not opted in by an admin

because of that I made the assumption that `isOptedIn()` will return `false` therefore it's like this opted out value is only meaningful when and admin has previously opted in the airnode. Maybe I'm the one missing something tho 😄

Reply:
haha, its more like this opted out value is a failsafe for the airnode to not mistakenly become part of the subscription model. 

Another use case is if they are already opted in, then they can choose to opt out temporarily and then join back in when they want.

Reply:
This should be implemented like
```
function setSelfOptOutStatus(bool _status)
        external
        override
    {
        airnodeSelfOptOutStatus[msg.sender] = _status;
        emit SetSelfOptOutStatus(msg.sender _status);
    }
```